### PR TITLE
Validation and global exception handler

### PR DESCRIPTION
**Pull request by BohdanZorii**

### CODE REVIEW

File: src/main/java/mate/zorii/bookstore/dto/BookDto.java [hunk 1]
```diff
@@ -1,15 +1,31 @@
...
+    @Size(max = 255, message = "Title must be less than 255 characters")
     private String title;
+    @NotBlank
+    @Size(max = 255, message = "Author must be less than 255 characters")
     private String author;
+    @NotBlank
+    @Pattern(regexp = "^(97(8|9))?\\d{9}(\\d|X)$", message = "Invalid ISBN format")
     private String isbn;
+    @NotNull
+    @Min(0)
```
Comment (MentorMateAcademy) [inline src/main/java/mate/zorii/bookstore/dto/BookDto.java:24]:
```suggestion
    @Positive
```

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                .toList();
+        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
+    }
+
+    private String getErrorMessage(ObjectError objectError) {
+        if (objectError instanceof FieldError fieldError) {
+            return fieldError.getField() + " " + fieldError.getDefaultMessage();
+        }
+        return objectError.getDefaultMessage();
+    }
```
Comment (sarakhmen) [inline src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java:29]: let's put private methods after public ones

File: src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
```
Comment (sarakhmen) [inline src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java:14]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/dto/CreateOrUpdateBookRequestDto.java [hunk 4]
```diff
@@ -1,14 +1,30 @@
...
     private String author;
+    @NotBlank
+    @Pattern(regexp = "^(97(8|9))?\\d{9}(\\d|X)$", message = "Invalid ISBN format")
     private String isbn;
+    @NotNull
+    @Min(0)
     private BigDecimal price;
+    @Size(max = 2000, message = "Description must be less than 2000 characters")
     private String description;
+
```
Comment (sarakhmen) [inline src/main/java/mate/zorii/bookstore/dto/CreateOrUpdateBookRequestDto.java:27]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/dto/BookDto.java [hunk 5]
```diff
@@ -1,15 +1,31 @@
...
     private String author;
+    @NotBlank
+    @Pattern(regexp = "^(97(8|9))?\\d{9}(\\d|X)$", message = "Invalid ISBN format")
     private String isbn;
+    @NotNull
+    @Positive
     private BigDecimal price;
+    @Size(max = 2000, message = "Description must be less than 2000 characters")
     private String description;
+
```
Comment (sarakhmen) [inline src/main/java/mate/zorii/bookstore/dto/BookDto.java:28]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/dto/CreateOrUpdateBookRequestDto.java [hunk 6]
```diff
@@ -1,14 +1,30 @@
...
+    @Size(max = 255, message = "Title must be less than 255 characters")
     private String title;
+    @NotBlank
+    @Size(max = 255, message = "Author must be less than 255 characters")
     private String author;
+    @NotBlank
+    @Pattern(regexp = "^(97(8|9))?\\d{9}(\\d|X)$", message = "Invalid ISBN format")
     private String isbn;
+    @NotNull
+    @Min(0)
```
Comment (sarakhmen) [inline src/main/java/mate/zorii/bookstore/dto/CreateOrUpdateBookRequestDto.java:23]: not fixed

**Review by sarakhmen — CHANGES_REQUESTED** [review]
let's improve your solution :)

File: pom.xml [hunk 7]
```diff
@@ -59,6 +59,10 @@
 			<artifactId>liquibase-core</artifactId>
 			<version>${liquibase.version}</version>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
```
Comment (Elena-Bruyako) [inline pom.xml:63]: Use spring-boot-starter-validation instead, which provides hibernate-validator also

File: src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleAllExceptions(Exception ex) {
+        return new ResponseEntity<>("An unexpected error occurred",
+                HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+
+    @ExceptionHandler(NullPointerException.class)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java:29]: remove

File: src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            MethodArgumentNotValidException ex) {
+        List<String> errors = ex.getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+        return new ResponseEntity<>(errors, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleAllExceptions(Exception ex) {
+        return new ResponseEntity<>("An unexpected error occurred",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/exception/GlobalExceptionHandler.java:25]: It's better provide more information about the exception. 
When you got this message, you won't understand what exactly happened

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution

**Comment by BohdanZorii** [comment]
accidentally commited to a wrong branch at first

File: src/main/java/mate/zorii/bookstore/dto/BookDto.java (hunk 5 above)
Comment (MentorMateAcademy) [reply src/main/java/mate/zorii/bookstore/dto/BookDto.java:28]: Not fixed

**Review by MentorMateAcademy — APPROVED** [review]

**Pull request merged**